### PR TITLE
Use Jupyterlab3/Federated Extension System

### PR DESCRIPTION
**Pull request by maartenbreddels**
This avoids users having to build/use nodejs for JLab3

### CODE REVIEW

**Comment by maartenbreddels** [comment]
@astrofrog could you take a look at this, the failure is because for some reason the pyproject.toml is not respected, it should install jupyter lab which is not part of the build process.

**Comment by astrofrog** [comment]
Yes will look shortly!

**Comment by astrofrog** [comment]
@maartenbreddels - I think the CI should be fixed now.

**Comment by astrofrog** [comment]
I also removed your workaround commit

**Comment by astrofrog** [comment]
Not sure what is going on with the coverage but I think we can ignore for now

**Comment by maartenbreddels** [comment]
@astrofrog the last two commits I don't understand:
 * ci: add setup.py to manifest, unclear why
 * ci: jupyter_packaging everywhere
 
I don't think they should be there, could you take a look?

**Comment by astrofrog** [comment]
Will take a look tomorrow morning!

**Comment by astrofrog** [comment]
@maartenbreddels - there were two separate issues:

* I think sdist is buggy in jupyter-packaging: https://github.com/jupyter/jupyter-packaging/pull/66 (I added a workaround here for now)
* There was a typo in the jupyter-packaging name in ``pyproject.toml``

Does this make sense?

**Comment by astrofrog** [comment]
Investigating a follow-up issue, think it should be fixable

**Comment by maartenbreddels** [comment]
Great, feel free to squash some commits if that cleans it up.

**Review by astrofrog — APPROVED** [review]
Looks good to me now! I think the setup.py could be simplified further as there is still some commented code.

**Comment by astrofrog** [comment]
Wait actually I think they just released a version of jupyter-packaging with the fox so I think we can remove the workaround and update the min pinned version in pyproject, will push a commit shortly

File: pyproject.toml [hunk 1]
```diff
@@ -0,0 +1,6 @@
+[build-system]
+requires = ["jupyter-packaging~=0.7.9",
```
Comment (astrofrog) [inline pyproject.toml:2]: Change to 0.7.12

File: setup.py [hunk 2]
```diff
@@ -1,165 +1,94 @@
...
-    def get_npm_name(self):
-        npmName = 'npm';
-        if platform.system() == 'Windows':
-            npmName = 'npm.cmd';
+# def get_data_files():
+#     return [
+#         ('share/jupyter/nbextensions/bqplot-image-gl', glob.glob('share/jupyter/nbextensions/bqplot-image-gl/*.js'),),
+#         ('share/jupyter/labextensions/bqplot-image-gl', glob.glob('share/jupyter/labextensions/bqplot-image-gl/*'),),
+#         ('etc/jupyter/nbconfig/notebook.d' ,['etc/jupyter/nbconfig/notebook.d/bqplot-image-gl.json'])
+#     ]
```
Comment (astrofrog) [inline setup.py:25]: Remove?

File: setup.py [hunk 3]
```diff
@@ -1,165 +1,94 @@
...
-            check_call([npmName, '--version'])
-            return True
-        except:
-            return False
+# def update_package_data(distribution):
+#     """update package_data to catch changes during setup"""
+#     build_py = distribution.get_command_obj('build_py')
+#     distribution.data_files = get_data_files()
+#     # re-init build_py options which load package_data
+#     build_py.finalize_options()
```
Comment (astrofrog) [inline setup.py:33]: Remove?

File: setup.py [hunk 4]
```diff
@@ -1,165 +1,94 @@
...
-        # update package data in case this created new files
-        update_package_data(self.distribution)
+js_command = combine_commands(
+    install_npm(js_dir, build_dir='share/jupyter/', source_dir='js/src', build_cmd='build'), ensure_targets(jstargets),
+)
 
-version_ns = {}
-with open(os.path.join(here, 'bqplot_image_gl', '_version.py')) as f:
-    exec(f.read(), {}, version_ns)
+# Adding "map_data" as package_data manually, this should not be needed because it's already
```
Comment (astrofrog) [inline setup.py:56]: Is this comment still relevant?

File: setup.py [hunk 5]
```diff
@@ -1,165 +1,94 @@
...
-            'bqplot_image_gl/static/index.js',
-            'bqplot_image_gl/static/index.js.map',
-        ],),
-        ('etc/jupyter/nbconfig/notebook.d' ,['etc/jupyter/nbconfig/notebook.d/bqplot-image-gl.json'])
-    ],
-    'install_requires': [
+# Backport of fix to sdist command in https://github.com/jupyter/jupyter-packaging/pull/66
+from setuptools.command.sdist import sdist
+from jupyter_packaging.setupbase import _wrap_command
+cmdclass['sdist'] = _wrap_command(['jsdeps'], sdist, strict=True)
```
Comment (astrofrog) [inline setup.py:68]: This workaround can be removed with jupyter packaging 0.7.12

**Review by astrofrog — CHANGES_REQUESTED** [review]
A few small things then I think this can be merged and released - thanks @maartenbreddels !

**Pull request merged**